checks/cs50/2017/x/finance/checks.py

```diff
@@ -1,13 +1,305 @@
+        if fields_filled != 3:
+            raise Error("Could not fill in username and/or password.")
+    
+        # search for register button
+        buttons = self.driver.find_elements_by_tag_name("button")
+        register_button = None
+        for button in buttons:
+            if button.text == "Register":
+                register_button = button
+        if register_button == None:
+            raise Error("Could not find register button.")
+        register_button.click()
+
+        if "Apology" in self.driver.title and outcome:
+            raise Error("Expected registration success, not rejection.")
+        elif "Apology" not in self.driver.title and not outcome:
+            raise Error("Expected registration rejection, not acceptance.")
+
+    def login(self, username, password):
+        """Checks that user can log in."""
+
+        # go to login page
+        self.driver.get("{}/login".format(Finance.BASE_PATH))
+        fields = self.driver.find_elements_by_tag_name("input")
+        fields_filled = 0
+        for field in fields:
+            name = field.get_attribute("name")
+            if name == "username":
+                fields_filled += 1
+                field.send_keys(username)
+            elif name == "password":
+                fields_filled += 1
+                field.send_keys(password)
+        if fields_filled != 2:
+            raise Error("Could not fill in username and/or password.")
+
+        # search for log in button
+        buttons = self.driver.find_elements_by_tag_name("button")
+        login_button = None
+        for button in buttons:
+            if button.text == "Log In":
+                login_button = button
+        if login_button == None:
+            raise Error("Could not find log in button.")
+        login_button.click()
+
+        if "Apology" in self.driver.title:
+            raise Error("Expected login to succeed, not fail.")
+
+    def quote(self, ticker, outcome, value=None):
+        """Checks that getting a quote results in desired outcome."""
+
+        # go to quote page
+        self.driver.get("{}/quote".format(Finance.BASE_PATH))
+        fields = self.driver.find_elements_by_tag_name("input")
     
+        # find field to input quote
+        quote_field = None
+        for field in fields:
+            name = field.get_attribute("name")
+            if name == "symbol":
+                quote_field = field
+        if quote_field == None:
+            raise Error("Could not input quote.")
+        quote_field.send_keys(ticker)
+
+        # search for quote button
+        buttons = self.driver.find_elements_by_tag_name("button")
+        quote_button = None
+        for button in buttons:
+            if button.text == "Quote":
+                quote_button = button
+        if quote_button == None:
+            raise Error("Could not find quote button.")
+        quote_button.click()
+
+        if "Apology" in self.driver.title and outcome:
```

Comment:
In my opinion, it really should be.

But in their words: "We will not be adding this feature to the WebDriver API as it falls outside of our current scope (emulating user actions)."